File: seminar_10/src/main/java/org/example/agregator/Product.java
/**
 * Класс Product представляет продукт, который может быть добавлен в заказ.
 * Этот класс содержит информацию о продукте, такую как идентификатор, название и стоимость.
 */
package org.example.agregator;

import java.util.Objects;

public class Product {
    private int id;
    private String name;
    private double price;

    /**
     * Конструктор класса Product.
     *
     * @param id    Идентификатор продукта.
     * @param name  Название продукта.
     * @param price Стоимость продукта.
     */
    public Product(int id, String name, double price) {
        this.id = id;
        this.name = name;
        this.price = price;
    }

    // Геттеры и сеттеры
    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public double getPrice() {
        return price;
    }

    public void setPrice(double price) {
        this.price = price;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Product product = (Product) o;
        return id == product.id &&
                Double.compare(product.price, price) == 0 &&
                Objects.equals(name, product.name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, name, price);
    }

    @Override
    public String toString() {
        return "Product{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", price=" + price +
                '}';
    }
}

File: app_with_ui/product.py
class Product:
    """ * Класс Product представляет товар, который может быть добавлен в заказ.
        * Этот класс содержит информацию о товаре, такую как идентификатор, название, стоимость и количество.
    """

    def __init__(self, id: int, name: str, price: float, count: int):
        """
        Инициализирует объект Product.

        Аргументы:
            id (int): Идентификатор товара.
            name (str): Название товара.
            price (float): Стоимость товара.
            count (int): Количество товара.
        """
        self.id = id
        self.name = name
        self.price = price
        self.count = count

    def __str__(self) -> str:
        """
        Возвращает строковое представление товара.

        Возвращает:
            str: Строковое представление товара.
        """
        return "id = {}, название = '{}', цена = {}, количество = {}".format(self.id, self.name, self.price, self.count)

    def __hash__(self) -> int:
        """
        Возвращает хэш-код товара.

        Возвращает:
            int: Хэш-код товара.
        """
        return hash((self.id, self.name, self.price, self.count))

    def __eq__(self, other) -> bool:
        """
        Проверяет равенство данного товара с другим товаром.

        Аргументы:
            other (Product): Другой товар для сравнения.

        Возвращает:
            bool: Результат проверки равенства товаров.
        """
        if isinstance(other, Product):
            return (self.id, self.name, self.price, self.count) == (other.id, other.name, other.price, self.count)
        return False
